sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -906,7 +906,19 @@ trait ScalaReflection extends Logging {
    * only defines a constructor via `apply` method.
    */
   private def getCompanionConstructor(tpe: Type): Symbol = {
-    tpe.typeSymbol.asClass.companion.asTerm.typeSignature.member(universe.TermName("apply"))
+    def throwUnsupportedOperation = {
+      throw new UnsupportedOperationException(s"Unable to find constructor for $tpe. " +
+        s"This could happen if $tpe is an interface, or a trait without companion object " +
+        "constructor.")
+    }
+    val companionSym = tpe.typeSymbol.asClass.companion match {
+      case NoSymbol => throwUnsupportedOperation
+      case sym => sym
+    }
```

Comment:
No big deal but I think you can just combine those two pattern matches:

```scala
tpe.typeSymbol.asClass.companion match {
  case NoSymbol =>
    throw new UnsupportedOperationException(s"Unable to find constructor for $tpe. " +
      s"This could happen if $tpe is an interface, or a trait without companion object " +
      "constructor.")
  case sym => sym.asTerm.typeSignature.member(universe.TermName("apply")) match {
    case NoSymbol =>
      throw new UnsupportedOperationException(s"Unable to find constructor for $tpe. " +
        s"This could happen if $tpe is an interface, or a trait with companion object " +
        "but that companion has no constructor and apply method.")
    case constructorSym => constructorSym
  }
}
```

Reply:
feel free to reword or rephrase ^.

Reply:
Updated to nested patter match, but retained the inner method to throw exception